### PR TITLE
Patterns: category page

### PR DESCRIPTION
**Pull request by nightnei**
Related to https://github.com/Automattic/dotcom-forge/issues/5849

## Proposed Changes
In the context of work related to the Patterns pages, this pull request includes the addition of markup for the `/patterns/:category` page.

## Testing Instructions
1. Open http://calypso.localhost:3000/patterns
2. Click on any pattern (e.g. `About`)
3. Asert that you are landed to `http://calypso.localhost:3000/patterns/about`
4. Assert that `About` category is active<br /> Screenshot 2024-03-06 at 15 40 02
5. Assert that in general you see the next page  <br /> Screenshot 2024-03-06 at 15 41 02
6. Assrt that navigation works well - click on items and scrolling left/right with clicks on arrows, also it's possible to scroll it with mouse-wheel
7. Assert that mobile version looks good and navigation doesn't have arrows but it's possible to scroll it with mouse-wheel <br /> Screenshot 2024-03-06 at 15 52 30

### CODE REVIEW

File: client/components/category-nav/index.tsx
Comment (stephanethomas) [inline client/components/category-nav/index.tsx]: We usually avoid abbreviations (see for example `NavigationHeader`), so I would recommend renaming that component to `CategoryNavigation`. Since we're introducing a new component that can be used in different places, what do you think of adding it to the [Calypso Docs](https://wpcalypso.wordpress.com/devdocs) to make it more discoverable? Finally, we could probably add a readme file that describes roughly what that component does (and maybe with a screenshot).

Comment (stephanethomas) [reply client/components/category-nav/index.tsx]: > We usually avoid abbreviations (see for example NavigationHeader)

Of course I've just found counter-examples 😅:

* `SectionNav` ([devdoc](https://wpcalypso.wordpress.com/devdocs/design/section-nav))
* `VerticalNav` ([devdoc](https://wpcalypso.wordpress.com/devdocs/design/vertical-nav))

So I'll let you decide 🙂. I wonder if we shouldn't also link all those components in each readme file so other developers are aware of those different options.

File: client/my-sites/patterns/index.tsx
Comment (stephanethomas) [inline client/my-sites/patterns/index.tsx]: This should be renamed.
```suggestion
type PatternsIndexProps = {
```

File: client/my-sites/patterns/index.tsx
Comment (stephanethomas) [inline client/my-sites/patterns/index.tsx]: We will most certainly have different titles for each page, for SEO reasons. So this should be moved (and duplicated) in each subcomponents.

Comment (stephanethomas) [reply client/my-sites/patterns/index.tsx]: Morever, given the very little code we have in that `PatternsIndex` component, I wonder if we should just ditch it.

File: client/my-sites/patterns/index.tsx
Comment (stephanethomas) [inline client/my-sites/patterns/index.tsx]: I'm not a big fan of the `Body` suffix, although I see where you're coming from.
```suggestion
				<PatternsHomePage />
			) : (
				<PatternsCategoryPage
					category={ category }
					isGridView={ isGridView }
					patternGallery={ PatternGallery }
				/>
```

That would also allow us to put them under a `pages` folder rather than a `bodies` folder (which sounds a bit strange)

File: client/my-sites/patterns/bodies/home/index.tsx
Comment (stephanethomas) [inline client/my-sites/patterns/bodies/home/index.tsx]: Why was the list of regular patterns removed?

File: client/my-sites/patterns/bodies/category/style.scss
Comment (stephanethomas) [inline client/my-sites/patterns/bodies/category/style.scss]:
```suggestion
.patterns-body-category__header {
```

File: client/my-sites/patterns/bodies/category/style.scss
Comment (stephanethomas) [inline client/my-sites/patterns/bodies/category/style.scss]:
```suggestion
	.patterns-body-category__title {
```

File: client/components/category-nav/index.tsx
Comment (nightnei) [reply client/components/category-nav/index.tsx]: >Since we're introducing a new component that can be used in different places, what do you think of adding it to the https://wpcalypso.wordpress.com/devdocs to make it more discoverable?

Yep, it's my plan for the follow up PR. Since we still don't have final designs, and additionally I have no idea how to add components to design sysyem, and also necessary to spend some time on description, screenshots etc. 

> We usually avoid abbreviations

On the one hand, it's better to keep consistency, but on the other hand, I agree - we have to avoid old mistakes with `Nav`, so I renamed it, thanks.

File: client/my-sites/patterns/index.tsx
Comment (nightnei) [reply client/my-sites/patterns/index.tsx]: No, since at least we should keep somewhere the common CSS (for header).
Also ofc `{ isLoggedIn && <UniversalNavbarHeader isLoggedIn /> }`.

File: client/my-sites/patterns/bodies/home/index.tsx
Comment (nightnei) [reply client/my-sites/patterns/bodies/home/index.tsx]: It's moved to `pages/category/index.tsx`. It was here temporarily before I created a page for it.

File: client/my-sites/patterns/index.tsx
Comment (stephanethomas) [reply client/my-sites/patterns/index.tsx]: And [a footer](https://github.com/Automattic/wp-calypso/pull/88245) soon 😀.

File: client/my-sites/patterns/pages/category/index.tsx
Comment (fredrikekelund) [inline client/my-sites/patterns/pages/category/index.tsx]:
```suggestion
				title={ category + ' patterns' }
```
Minor, and I realize that this isn't final regardless, but let's skip `page` here.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]: Let's just call `usePatternCategories` inside this component instead.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]:
```suggestion
		listRef.current.addEventListener( 'scroll', checkScrollArrows, { passive: true } );

		return () => {
			listRef.current.removeEventListener( 'scroll', checkScrollArrows );
```
Referencing `listRef.current` directly is just my personal preference (I see no reason to use the `list` variable), but `{ passive: true }` should definitely be moved from the `removeEventListener` call to `addEventListener`

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]: I'd consider moving this outside the `CategoryNavigation` scope and passing `listRef.current` as an argument.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]: It feels like we're registering the same event listener twice with `onScroll` here and the `useEffect` callback. Do we really need both?

File: client/components/category-navigation/style.scss
Comment (fredrikekelund) [inline client/components/category-navigation/style.scss]:
```suggestion
	flex-shrink: 0;
```
I think this would communicate the intention more clearly.

**Review by fredrikekelund — COMMENTED** [review]
This is looking great overall 👍 I left some initial comments. Will continue digging some more too

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]: I think we could define this inside the component too. No real upside to passing it as a prop (right?)

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]: Do we expect to pass `children` to this component? Feels like this might be a leftover?

File: client/components/category-navigation/style.scss
Comment (fredrikekelund) [inline client/components/category-navigation/style.scss]:
```suggestion
```
It doesn't look like this makes any difference.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]:
```suggestion
							<>
								{ buttons.map( ( button ) => (
									<a
										key={ button.label }
										href={ button.link }
										className="category-navigation__button"
									>
										{ button.icon && <img src={ button.icon } alt={ button.label } /> }
										{ button.label }
									</a>
								) ) }
								<div className="category-navigation__button-divider" />
							</>
```
Optional, but the `category-navigation__buttons` class name is somewhat confusing, and I think we could probably use a separator element just as well as an inner wrapper.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [reply client/components/category-navigation/index.tsx]: I didn't realize this was meant to be a generalized component. Hmm…

File: client/my-sites/patterns/pages/category/index.tsx
Comment (stephanethomas) [reply client/my-sites/patterns/pages/category/index.tsx]: Something to keep in mind for the future: we won't be able to concatenate content like this when we localize that title. We can certainly use a placeholder but if we want to optimize that page for search engines we will probably have to follow the approach used in the `Themes` page where a different title is defined [statically](https://github.com/Automattic/wp-calypso/blob/975b3c2a995d3f39bd8e1904d1b90c0276133116/client/my-sites/themes/use-theme-showcase-logged-out-seo-content.js#L26) for each category.

Comment (nightnei) [reply client/my-sites/patterns/pages/category/index.tsx]: Thanks @stephanethomas for point to nice example.
The current approach is just a dummy copy, to render different headers.

File: client/my-sites/patterns/index.tsx
Comment (fredrikekelund) [inline client/my-sites/patterns/index.tsx]:
```suggestion
import { PatternsCategoryPage } from 'calypso/my-sites/patterns/pages/category';
import { PatternsHomePage } from 'calypso/my-sites/patterns/pages/home';
```

File: client/components/category-navigation/index.tsx
Comment (nightnei) [reply client/components/category-navigation/index.tsx]: It's not final version, I will adjust it to be used as a common component for different places.

File: client/components/category-nav/index.tsx
Comment (fredrikekelund) [reply client/components/category-nav/index.tsx]: Given that we plan to reuse this nav, it does make sense to implement it as a generalized component. I would consider tweaking the name to something like `CategoryPillNavigation`, which I think is more descriptive.

File: client/components/category-navigation/index.tsx
Comment (nightnei) [reply client/components/category-navigation/index.tsx]: > Referencing listRef.current directly is just my personal preference (I see no reason to use the list variable)

Screenshot 2024-03-07 at 13 23 23

> but { passive: true } should definitely be moved from the removeEventListener call to addEventListener

Oh, sharp eyes Fredrik, thanks. It's indeed should be there, just typo.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [reply client/components/category-navigation/index.tsx]: It's my bad for not realizing this is a generalized component. However, it looks like `Discover` and `All categories` will also be reused (p1709817801121679/1709817468.279259-slack-C06ELHR6L9J) in the other places where we plan to use this component. Maybe we shouldn't pass those as props after all

Comment (fredrikekelund) [reply client/components/category-navigation/index.tsx]: `All categories` is meant to be a dropdown and not a link in and of itself, which might also support hard coding it instead of passing it as a prop BnTfb6hzKcRTb2LBtuPwKe-fi-246:42348#727280144

File: client/components/category-navigation/index.tsx
Comment (nightnei) [reply client/components/category-navigation/index.tsx]: Sorry, I didn't get your point and reason.

My thoughts - we are creating a common component that should behave the same way everywhere for consistency. So, this should definitely be here to avoid different animations on different pages.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [reply client/components/category-navigation/index.tsx]: This still applies, even if I didn't realize at first that this was meant to be a generalized component

File: client/my-sites/patterns/pages/category/index.tsx
Comment (fredrikekelund) [reply client/my-sites/patterns/pages/category/index.tsx]: I was thinking about how we should handle translations. Thanks for digging up that reference, @stephanethomas 👍

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [reply client/components/category-navigation/index.tsx]: This is what I had in mind 33c1d-pb/#js

This is optional, and I don't think we can realistically talk about performance benefits or anything like that with with my suggested approach, it's more a matter of code organization.

`scrollTo` doesn't _have to_ live in the component scope, it could easily live outside, which would make the component a little bit more compact (and easier to read).

Comment (nightnei) [reply client/components/category-navigation/index.tsx]: Oh, got it. My thoguhts are:
1) If we move `scrollTo`, then we should also move `checkScrollArrows` outside.
2) If we don't have a dependency on the DOM node - I would totally agree, but in our case - we have, so I would keep it inside.

File: client/components/category-navigation/index.tsx
Comment (nightnei) [reply client/components/category-navigation/index.tsx]: >It's my bad for not realizing this is a generalized component. However, it looks like Discover and All categories will also be reused (p1709817801121679/1709817468.279259-slack-C06ELHR6L9J) in the other places where we plan to use this component. Maybe we shouldn't pass those as props after all

I am not sure about the right approach. I suppose that we have different handlers, also I don't want to make the buttons hardcoded, to provide flexibility for any future cases. I propose to keep it as is, and we will see later, when the designs are completely ready etc.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [reply client/components/category-navigation/index.tsx]: `checkScrollArrows` has a dependency on state, so I don't think it makes as much sense to move that outside `CategoryNavigation`.

With `scrollTo`, the function isn't as closely tied to any particular ref/state, and it's simple to pass an `element` argument, which is why I think it might make sense in that case.

But if you prefer to keep it as is, then that's fine by me!

File: client/components/category-navigation/index.tsx
Comment (nightnei) [reply client/components/category-navigation/index.tsx]: >All categories is meant to be a dropdown and not a link in and of itself, which might also support hard coding it instead of passing it as a prop [BnTfb6hzKcRTb2LBtuPwKe-fi-246:42348#727280144](https://href.li/?https://www.figma.com/file/BnTfb6hzKcRTb2LBtuPwKe/?node-id=246:42348#727280144)

Hm, maybe I agree with you. A lot of questions... 😅 I propose to keep it also as is, to unblock Stephane and have some basic markup, and I will adjust it in the followup PR. I think that this way will speed up other features/functionality.

File: client/components/category-navigation/index.tsx
Comment (nightnei) [reply client/components/category-navigation/index.tsx]: It will be necessary to provive inputs and other components, take a look at examples for `/themes` p9Jlb4-aX8-p2

File: client/components/category-navigation/style.scss
Comment (nightnei) [reply client/components/category-navigation/style.scss]: It will make difference, since there is `children`. I just prepare everything and keep in mind for future cases with `/themes` and `/blog`.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [reply client/components/category-navigation/index.tsx]: Sure! We can definitely return to this in a later PR

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [reply client/components/category-navigation/index.tsx]: I see! Two questions come to mind when looking at those drafts:
- Do those components really need to live inside `CategoryNavigation`, or could they be siblings?
- If they do need to live inside `CategoryNavigation`, would it perhaps make more sense to pass them as a prop?

I would argue that we should return to this once we get around to implementing this component on `/themes` and not include the `children` prop until then.

Comment (nightnei) [reply client/components/category-navigation/index.tsx]: > Do those components really need to live inside CategoryNavigation, or could they be siblings?

Good question. I think I agree with you.

> If they do need to live inside CategoryNavigation, would it perhaps make more sense to pass them as a prop?

I am not sure that it's good idea, but anyway, now, after thinking a few more times - I totally agree with this: `I would argue that we should return to this once we get around to implementing this component on /themes and not include the children prop until then.`

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]:
```suggestion
								<LocalizedLink
									key={ button.label }
									href={ button.link }
									className="category-navigation__button"
								>
									{ button.icon && <img src={ button.icon } alt={ button.label } /> }
									{ button.label }
								</LocalizedLink>
```
We should either find a way to use `LocalizedLink` from `client/my-sites/patterns/components/localized-link.tsx` in here, or we should implement the corresponding logic when generating the `href`s in `client/my-sites/patterns/pages/category/index.tsx`.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]: Same remark here about using `LocalizedLink`, or passing correctly localized `href`s through the props to this component.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]:
```suggestion
		if ( ! listRef.current ) {
			return;
		}

		const { scrollLeft, scrollWidth, clientWidth } = listRef.current;
```
Nit, but I see no reason to abstract `listRef.current` into a temporary variable.

File: client/components/category-navigation/index.tsx
Comment (fredrikekelund) [inline client/components/category-navigation/index.tsx]:
```suggestion
export const CategoryPillNavigation = ( { buttons, list, selectedCategory }: CategoryPillNavigationProps ) => {
```
I mentioned this here (https://github.com/Automattic/wp-calypso/pull/88244#discussion_r1516143446) as well, but WDYT about tweaking the name of this component like this? `CategoryNavigation` seems quite generic, but including "pill" hints at what the component looks like.

File: client/my-sites/patterns/hooks/use-patterns.ts
Comment (fredrikekelund) [inline client/my-sites/patterns/hooks/use-patterns.ts:20]: How come you removed this? The idea was not to fetch patterns on the `/patterns` home page because we don't render any pattern previews there.

File: client/my-sites/patterns/index.tsx
Comment (fredrikekelund) [inline client/my-sites/patterns/index.tsx]: Instead of adding this component, we could also implement the same logic in `renderPatterns` in `index.node` and `index.web` respectively. Do you think this approach is better?

File: client/my-sites/patterns/pages/category/index.tsx
Comment (fredrikekelund) [inline client/my-sites/patterns/pages/category/index.tsx]:
```suggestion
							link: '/patterns',
```
Let's make this a link back to the `/patterns` home page. 

Again, if we don't add `LocalizedLink` in `CategoryNavigation`, then we need to remember to localize the link here.

File: client/my-sites/patterns/pages/category/style.scss
Comment (fredrikekelund) [inline client/my-sites/patterns/pages/category/style.scss]:
```suggestion
.is-section-patterns .category-navigation {
```
We should qualify the `.category-navigation` selector somehow.

**Review by fredrikekelund — COMMENTED** [review]
This is shaping up nicely, but there are a few things we should still address before landing!

File: client/components/category-nav/index.tsx
Comment (nightnei) [reply client/components/category-nav/index.tsx]: I don't mind such a name, renamed

File: client/components/category-navigation/index.tsx
Comment (nightnei) [reply client/components/category-navigation/index.tsx]: I don't like names based on UI, since UI can be changed. But I agree that we should have less generic name, since we already have similar genric components. So I am ok with the proposed name.

File: client/my-sites/patterns/hooks/use-patterns.ts
Comment (nightnei) [reply client/my-sites/patterns/hooks/use-patterns.ts:20]: I did it intentionally, since it's redundant ATM - we always have `category` https://github.com/Automattic/wp-calypso/pull/88244/files#diff-1daf550a7455ffaec1000ea6febdd24fa3ef5c6ed6f5e04b53c36a2eee1d64d1R29-R33

File: client/my-sites/patterns/index.tsx
Comment (nightnei) [reply client/my-sites/patterns/index.tsx]: I think that this approach is better:
1) There is an answer above https://github.com/Automattic/wp-calypso/pull/88244#discussion_r1514999285
2) Also, with this approach we avoid duplications
3) With separate file we decouple - logic (index.node and index.web) from rendering (index)

File: client/my-sites/patterns/hooks/use-patterns.ts
Comment (fredrikekelund) [reply client/my-sites/patterns/hooks/use-patterns.ts:20]: Right 😄

File: client/my-sites/patterns/index.tsx
Comment (fredrikekelund) [reply client/my-sites/patterns/index.tsx]: Fair enough! Could we maybe rename this to `PatternsWrapper` or something, though? `PatternsIndex ` sounds like this might be the home page for patterns

Comment (nightnei) [reply client/my-sites/patterns/index.tsx]: >PatternsIndex  sounds like this might be the home page for patterns

Of course, it's personal, but it doesn't sound to me this way. And when I started my journey in development many years ago - exactly PHP examples showed me the difference - typically index.php was used as a layout - common header and footer plus some body (based on a route). So `index` for me is entry point, when `home` is mostly about `body` inside `index`.

ChatGPT says:
```
Differences
Functionality Scope: index.php is often used for initialization and routing, wrapper.php for shared functionality or layout wrapping, and home.php for defining the homepage content.
```

So maybe we could use `wrapper`, but exactly our case looks like "typical index.php" file to me :) 

WDYT?

Comment (fredrikekelund) [reply client/my-sites/patterns/index.tsx]: `index` to me is the default lookup location in a folder, i.e., the home page. [This Wikipedia article](https://en.wikipedia.org/wiki/Web_server_directory_index) puts it well:
> When an HTTP client (generally a web browser) requests a URL that points to a directory structure instead of an actual web page within the directory structure, the web server will generally serve a default page, which is often referred to as a main or "index" page.

This is also true for import paths in Calypso, which is why `client/my-sites/patterns/components/header/index.tsx` is looked up when we call `import { PatternsHeader } from 'calypso/my-sites/patterns/components/header'`, despite there not being an `index.tsx` in the import path.

Comment (nightnei) [reply client/my-sites/patterns/index.tsx]: 👍 Renamed

**Review by fredrikekelund — APPROVED** [review]
LGTM 👍 Thanks for working through this. Let's get this landed and then keep iterating!

**Pull request merged**
